clients/src/main/java/org/apache/kafka/common/utils/CollectionUtils.java

```diff
@@ -49,11 +50,22 @@ private CollectionUtils() {}
     }
 
     /**
+     * @deprecated Replaced by {@link #groupDataByTopic(Collection)}
      * group partitions by topic
      * @param partitions
      * @return partitions per topic
      */
+    @Deprecated
     public static Map<String, List<Integer>> groupDataByTopic(List<TopicPartition> partitions) {
+        return groupDataByTopic((Collection) partitions);
+    }
+
+    /**
+     * group partitions by topic
+     * @param partitions
+     * @return partitions per topic
+     */
+    public static Map<String, List<Integer>> groupDataByTopic(Collection<TopicPartition> partitions) {
         Map<String, List<Integer>> partitionsByTopic = new HashMap<>();
```

Comment:
Is it really necessary to keep around both functions?  It seems like this utility function is only supposed to be used by Kafka, not by end-users.

Reply:
My intention was to retain binary compatibility. Since it's not in an internal package how would an end-user know that they weren't supposed to use this API?

Reply:
They should know because there isn't an interface annotation that declares it a public API.

There's only 5 files in `./clients/src/main/java/org/apache/kafka/common/internals/`.  If you take the position that everything else is a public API, we're in deep trouble!

Reply:
We publish javadoc for public APIs. Everything else is not public. This is a bit opaque and it would be better to have annotations as @cmccabe suggests.